clevertap-core/src/main/java/com/clevertap/android/sdk/network/api/CtApi.kt

```diff
@@ -41,74 +46,192 @@ internal class CtApi(
+            createRequest(
+                baseUrl = getActualDomain(isViewedEvent = false) ?: defaultDomain,
+                relativeUrl = "defineTemplates",
+                body = body.toString()
+            )
         )
 
-    fun getActualDomain(useSpikyDomain: Boolean): String? {
+    fun getActualDomain(isViewedEvent: Boolean): String? {
         return when {
             !region.isNullOrBlank() -> {
-                val regionSuffix = if (useSpikyDomain) spikyRegionSuffix else ""
-                "$region${regionSuffix}.$defaultDomain"
+                buildString {
+                    append(region)
+                    append(
+                        if (isViewedEvent) {
+                            spikyRegionSuffix
+                        } else {
+                            ""
+                        }
+                    )
+                    append(".")
+                    append(defaultDomain)
+                }
             }
 
-            !useSpikyDomain && !proxyDomain.isNullOrBlank() -> {
+            !isViewedEvent && !proxyDomain.isNullOrBlank() -> {
                 proxyDomain
             }
 
-            useSpikyDomain && !spikyProxyDomain.isNullOrBlank() -> {
+            isViewedEvent && !spikyProxyDomain.isNullOrBlank() -> {
                 spikyProxyDomain
             }
 
-            else -> if (useSpikyDomain) {
-                spikyDomain
+            else -> if (isViewedEvent) {
+                cachedSpikyDomain
             } else {
-                domain
+                cachedDomain
+            }
+        }
+    }
+
+    fun getHandshakeDomain(isViewedEvent: Boolean) : String {
+        if (region.isNotNullAndBlank()) {
+            return buildString {
+                append(region)
+                append(
+                    if (isViewedEvent) {
+                        spikyRegionSuffix
+                    } else {
+                        ""
+                    }
+                )
+                append(".")
+                append(defaultDomain)
+            }
+        }
+
+        if (isViewedEvent) {
+            if (spikyProxyDomain.isNotNullAndBlank()) {
+                return spikyProxyDomain!!
+            }
+        } else {
+            if (proxyDomain.isNotNullAndBlank()) {
+                return proxyDomain!!
+            }
+        }
+
+        if (customHandshakeDomain.isNotNullAndBlank()) {
+            return customHandshakeDomain!!
+        }
+
+        if (isViewedEvent) {
+            if (cachedSpikyDomain.isNotNullAndBlank()) {
+                return cachedSpikyDomain!!
+            }
+        } else {
+            if (cachedDomain.isNotNullAndBlank()) {
+                return cachedDomain!!
+            }
+        }
+
+        return defaultDomain
+    }
+
+    fun needsHandshake(
+        isViewedEvent: Boolean,
+        clearCachedHandshakeDomain: () -> Unit = {}
+    ) : Boolean {
+
+        if (region.isNotNullAndBlank()) {
```

Comment:
Improvements by GPT 
```
    if (region.isNotNullAndBlank()) {
        clearCachedHandshakeDomain()
        return false
    }

    val proxyDomain = if (isViewedEvent) spikyProxyDomain else proxyDomain
    if (proxyDomain.isNotNullAndBlank()) {
        return false
    }

    if (customHandshakeDomain.isNotNullAndBlank()) {
        return cachedHandshakeDomain.isNullOrBlank()
    }

    clearCachedHandshakeDomain()

    return if (isViewedEvent) {
        cachedSpikyDomain.isNullOrBlank()
    } else {
        cachedDomain.isNullOrBlank()
    }
```

Reply:
Do take a look at changes. Made some changes, also logic is slightly changed.